File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        deque<TreeNode*> stack1;
        if (root) {  // if not None
            stack1.push_back(root);
        }
        deque<TreeNode*> stack2;
        vector<int> res;  // result

        TreeNode* rightmost;
        while (!stack1.empty()) {
            // rightmost node
            rightmost = stack1.back();
            // bfs from left to right
            while (!stack1.empty()) {
                if (stack1.front()->left) {
                    stack2.push_back(stack1.front()->left);
                }
                if (stack1.front()->right) {
                    stack2.push_back(stack1.front()->right);
                }
                stack1.pop_front();
            }
            // extract rightmost element of stack1 (= to node here!)
            res.push_back(rightmost->val);
            // stack1 is stack2, stack1 become empty
            swap(stack1, stack2);
        }
        return res;
    }
};

File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        self.stack1 = []
        if root:  # if not None
            self.stack1.append(root)
        self.stack2 = []
        res = []  # result
        while self.stack1:
            # bfs
            for node in self.stack1:
                if node.left:
                    self.stack2.append(node.left)
                if node.right:
                    self.stack2.append(node.right)
            # extract rightmost element of stack1
            res.append(self.stack1[-1].val)
            # stack1 is stack2
            self.stack1 = self.stack2[:]
            self.stack2.clear()
        return res
